### ISSUE TITLE
[css-view-transitions-2] Elements with content-visibility in new Document

### ISSUE THREAD

**Issue by khushalsagar** [w3c/csswg-drafts#10773]
If an element in the new Document for a cross-document transition has `content-visibility: auto`, it's `view-transition-name` is skipped. This is because the algorithm to look for names runs before the first rendering loop on the new Document. And the initial state of elements with `content-visibility: auto` is computed as a part of update the rendering [here](https://html.spec.whatwg.org/#update-the-rendering:~:text=Let%20hadInitialVisibleContentVisibilityDetermination%20be%20false.).

The simple fix would be to force the determination of this state if a Document has an incoming cross-document ViewTransition.

@vmpstr @argyleink @noamr @nt1m

**Comment by nt1m** [comment]
cc @mattwoodrow

**Comment by nt1m** [comment]
> The simple fix would be to force the determination of this state if a Document has an incoming cross-document ViewTransition.

Makes sense but seems a bit weird to have a one-off exception

**Comment by khushalsagar** [comment]
> I guess it works, but seems a bit weird to have a one-off exception

Not exactly one-off. `content-visibility` is subtle this way. There's other cases like `getComputedStyle` which force computing style for an element outside the rendering loop. In this case we don't need to force computing style, since VT spec already says that elements which aren't relevant to the user can be ignored. But we need to know which elements are relevant to the user.

That said, happy to hear if you have any other suggestions.

Early on we considered running the [capture the new state](https://drafts.csswg.org/css-view-transitions-1/#capture-new-state-algorithm) algorithm within update the rendering loop that would have avoided it. But we needed to give authors a script hook to set up web animations targeting pseudo-elements after this step. And more script hooks in the later stages of update the rendering become a pain to maintain. There's already a loop there which has to run resize observer and content visibility state tracking multiple times until it settles.

**Comment by emilio** [comment]
We could also assume `content-visibility: auto` elements are relevant for the purposes of view-transition names. We do that for find in page and scroll-into-view already, correct?

**Comment by noamr** [comment]
I think we should update the https://drafts.csswg.org/css-contain-2/#relevant-to-the-user view-transition condition to include elements with a `view-transition-name` when there is an inbound transition

**Comment by emilio** [comment]
That on its own doesn't work tho, right? (Because of the timing issue described above)

**Comment by khushalsagar** [comment]
> All of this happens before the content-visibility loop.

That's what's causing this problem and why it particularly affects only new DOM capture. For old DOM capture, looking for names happens after this step.

> We could also assume content-visibility: auto elements are relevant for the purposes of view-transition names. We do that for find in page and scroll-into-view already, correct?

The problem with this is that it makes `content-visibility` somewhat irrelevant as an optimization if you add VT to your page. One of its goals was to make FCP faster by avoiding style/layout of content which is far offscreen. If we force it to be relevant on the first frame then that gain is lost.

The issue where we resolved on this behaviour also concluded that there's no user benefit to forcing this content to be relevant. There's no use-case where far offscreen content needs to participate in the transition. We just don't know the position of these nodes until the first layout to determine whether they should be locked or not. That's why the ideal fix would be to make that determination before we look for names in the new DOM.

Tangential question, "scroll-into-view" reminded me of "scroll-to-text". I suppose it's like having "find-in-page" automatically run on page load where we're forcing nodes to become relevant. I'd still say there's a user benefit/need to do that here but not for VT.

**Comment by noamr** [comment]
> > All of this happens before the content-visibility loop.
> 
> That's what's causing this problem and why it particularly affects only new DOM capture. For old DOM capture, looking for names happens after this step.

Ah so the purpose of the current condition in css-contain is to keep elements relevant *after* the captures I guess?

**Comment by khushalsagar** [comment]
Yup. We want to keep it relevant once we've decided to capture. But won't force anything to become relevant to look for names.

**Comment by noamr** [comment]
I wonder if the solution would be to not activate the transition synchronously in `pagereveal`, but rather do so at the usual "run pending transition operations" step that comes later, but I haven't thought through the implications.

**Comment by vmpstr** [comment]
The problem is that we need to know whether there will be a valid viewTransition object at pagereveal time, which necessitates getting view transition names -- duplicates / conflicts would cause a null viewTransition object.

I do think that perhaps this is only a problem for cross doc view transitions, since looking at the timing of same document update callback etc, we should already be processing c-v auto visibility changes before detecting the view transition names.

**Comment by noamr** [comment]
> The problem is that we need to know whether there will be a valid viewTransition object at pagereveal time, which necessitates getting view transition names -- duplicates / conflicts would cause a null viewTransition object.

Maybe that's not necessary? We can have a `viewTransition` object and reject its `ready` promise if there are duplicates, like we do in the SPA case.

**Comment by noamr** [comment]
A few notes from internal sync:
- This is not only a cross-document issue. Also when elements are added during the update callback of a same-document transition, they would need their initial content-visibility determination before capturing the new state.
- In chromium there is a "forced lifecycle" step before capturing the new state that does this.
- Other things in the rendering lifecycle can affect the new state capture, e.g. the right elements might not be there at all before  determining the size via a `ResizeObserver`.

I see two options on how to resolve this:

## 1. "Forced" style and layout + content visibility determination before capturing the new state

This would not be a cross-document specific thing, but rather a view-transition "capture the new state" thing - right before capturing the new state, we would calculate the style, run the layout, and also make the initial determination of content visibility.

## 2. Async capturing of new state, during the style/layout loop

Instead of capturing the new state synchronously as a result of resolving the update callback (same-document) or revealing a new page (cross-document), we would queue that, and run it at the style/layout/`ResizeObserver` loop's first iteration, after determining the initial content-visibility state. Note that that capture also resolves the `ready` promise of the `ViewTransition`.

Though (2) feels perhaps more correct, it would also be a breaking change, as it would inverse the order when it comes to `requestAnimationFrame`, which would now be called mid-transition, before the `ready` promise of the `ViewTransition` object.

@emilio @vmpstr @nt1m thoughts?

**Comment by vmpstr** [comment]
One of the use cases of the `ready` promise is for authors to synchronize their own animations with the view transition one. If the proposal is to resolve the promise during the lifecycle, then I believe naively that script would run after the paint for the frame has already been produced, which means there would be one frame before this synchronization can happen.

**Comment by noamr** [comment]
OK, discussed this at internal sync.
Proposal is to perform initial content visibility state if needed as we're iterating over the element tree when capturing the new state. So when we walk the tree, and run into an element with `content-visibility: auto` whose [proximity to the viewport](https://drafts.csswg.org/css-contain/#proximity-to-the-viewport) is not determined and it is not [relevant to the user](https://drafts.csswg.org/css-contain/#relevant-to-the-user), we determine [proximity to the viewport](https://drafts.csswg.org/css-contain/#proximity-to-the-viewport) for that element.

From a web-observable perspective it's roughly equivalent to what was proposed before, but it's clearer here that this is specifically for making sure this state is determined before checking for view-transition elgibility.
